File: subarray_sum_equals_k.py
#Time Complexity = O(n)
#Space Complexity = O(1)
#Leetcode = yes
# approach = we we calculate running sum and every index and store running sum and its count in hashmap and whenever we find that rsum-k is in hahshmap we will add its count to our current count and if rsum - k is not in hashmap then we will add its count. 
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        runningSum = 0
        HashMap = {0: 1}  # rsum : rsum_count
        count = 0

        for num in nums:
            runningSum += num
            if runningSum - k in HashMap:
                count += HashMap[runningSum - k]
            HashMap[runningSum] = HashMap.get(runningSum, 0) + 1

        return count

File: SubarraySumEqualsK.cs
//Algo: using running sum of array
//TC:O(N)
//SC:O(N) HashMap length

public class Solution {
    public int SubarraySum(int[] nums, int k) {
        int rSum = 0;
        int count = 0;
        Dictionary<int,int> map = new Dictionary<int,int>();
        map.Add(0,1);
        
        for(int i=0;i<nums.Length;i++){
            rSum+=nums[i];
            
            int targetDiff = rSum-k;
            
            if(map.ContainsKey(targetDiff)){
                count+=map[targetDiff];
            }
            
            if(!map.ContainsKey(rSum)){
                map.Add(rSum,1);
            }else{
                map[rSum]= map[rSum]+1;
            }
        }
        
        return count;
    }
}